src/dataflow/src/source/s3.rs

```diff
@@ -515,27 +552,57 @@ enum Sent {
     SenderClosed,
 }
 
+#[derive(Debug)]
+enum S3Error {
+    BodyMissing(String),
+    ClientConstructionFailed(anyhow::Error),
+    Decode(String, std::io::Error),
+    GetObjectError(RusotoError<rusoto_s3::GetObjectError>),
+    ListObjectsFailed(RusotoError<rusoto_s3::ListObjectsV2Error>),
+    Read(std::io::Error),
+    RetryFailed,
+}
+
+impl std::fmt::Display for S3Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            S3Error::BodyMissing(body) => write!(f, "Get object response had no body: {}", body),
+            S3Error::ClientConstructionFailed(err) => err.fmt(f),
+            S3Error::Decode(key, err) => {
+                write!(f, "Failed to decode object {} using gzip: {}", key, err)
+            }
+            S3Error::GetObjectError(err) => err.fmt(f),
+            S3Error::ListObjectsFailed(err) => err.fmt(f),
+            S3Error::Read(err) => err.fmt(f),
+            S3Error::RetryFailed => write!(f, "Retry failed to produce result"),
+        }
+    }
+}
+
+type S3Result<R> = Result<R, S3Error>;
+
 async fn download_object(
-    tx: &SyncSender<anyhow::Result<InternalMessage>>,
+    tx: &SyncSender<S3Result<InternalMessage>>,
     activator: &SyncActivator,
     client: &S3Client,
-    bucket: String,
-    key: String,
+    bucket: &str,
+    key: &str,
     compression: Compression,
-) -> Option<DownloadMetricUpdate> {
+) -> (bool, Option<DownloadMetricUpdate>) {
```

Comment:
Could you either document what this boolean return value means in a docstring, or, possibly better, convert the `Option` into a custom three- or four-branched enum?